pkg/buildchange/change_logger.go

```diff
@@ -0,0 +1,92 @@
+package buildchange
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/pivotal/kpack/pkg/differ"
+)
+
+const differPrefix = "\t"
+
+func Log(logger *log.Logger, changesStr string) error {
+	return NewChangeLogger(logger, changesStr).Log()
+}
+
+func NewChangeLogger(logger *log.Logger, changesStr string) *changeLogger {
+	options := differ.DefaultOptions()
+	options.Prefix = differPrefix
+
+	return &changeLogger{
+		logger:     logger,
+		changesStr: changesStr,
+		differ:     differ.NewDiffer(options),
+	}
+}
+
+type changeLogger struct {
+	logger     *log.Logger
+	changesStr string
+
+	differ  differ.Differ
+	reasons []string
+	changes []GenericChange
+}
+
+func (c *changeLogger) Log() error {
+	if c.changesStr == "" {
+		return nil
+	}
+
+	if err := c.parseChanges(); err != nil {
+		return errors.Wrapf(err, "error parsing build changes JSON string '%s'", c.changesStr)
+	}
+	c.parseReasons()
+
+	c.logReasons()
+	return c.logChanges()
+}
+
+func (c *changeLogger) parseChanges() error {
+	c.changes = []GenericChange{}
+	return json.Unmarshal([]byte(c.changesStr), &c.changes)
+}
+
+func (c *changeLogger) parseReasons() {
+	c.reasons = make([]string, len(c.changes))
+	for i, change := range c.changes {
+		c.reasons[i] = change.Reason
+	}
+}
+
+func (c *changeLogger) logReasons() {
+	reasons := strings.Join(c.reasons, reasonsSeparator)
+	c.logger.Printf("Build reason(s): %s\n", reasons)
+}
+
+func (c *changeLogger) logChanges() error {
+	for _, change := range c.changes {
+		if change.Reason == "TRIGGER" {
```

Comment:
Is this Trigger check needed?

Reply:
> Is this Trigger check needed?

The TRIGGER check was done to match @sampeinado 's requirement on how to display aTRIGGER change
Happy to remove the conditional check If displaying a TRIGGER change as a diff like the other changes is acceptable

Reply:
@matthewmcnew and @sampeinado : please let me know how to proceed on this. I don't feel strongly about going either way.

Reply:
Hmm. It seems like we should try to avoid treating build changes individually if that is at all possible. 

Would it be possible to just drop the phrase `change:`?

Reply:
So, that could look like: 

```
Build reason(s): TRIGGER,BUILDPACK
TRIGGER: A new build was manually triggered on Mon, 23 Nov 2020 16:16:26 -0500
BUILDPACK:
	- tanzu-buildpacks/go-build	0.0.23
...

```

Reply:
Fine.. will treat them all the same. It won't look exactly like how you listed. more like this
```
```

Reply:
Fine.. will treat them all the same. It won't look exactly like how you listed. more like this
```
Build reason(s): TRIGGER,BUILDPACK
TRIGGER:
        + A new build was manually triggered on Mon, 07 Dec 2020 18:29:42 -0500
BUILDPACK:
	- id: tanzu-buildpacks/go-build
	-   version: 0.0.23
	+   version: 0.0.22
	- id: tanzu-buildpacks/go-dist
	-   version: 0.1.2
	+   version: 0.1.0
	- id: tanzu-buildpacks/go-mod-vendor
	-   version: 0.0.25
	+   version: 0.0.24
```

just a general observation... bending/altering design requirements for avoiding special handling (which is not a big deal) in implementation is a questionable approach.

However, I do agree that In this scenario, the output does look good even if the changes are handled the same way.

Reply:
resolved by https://github.com/pivotal/kpack/commit/be3ad6dd90eb514a1ce408e7c72bb7f649a74ff2